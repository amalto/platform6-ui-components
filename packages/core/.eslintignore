dist/
loader/

src/components.d.ts