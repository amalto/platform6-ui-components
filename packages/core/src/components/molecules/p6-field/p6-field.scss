$label-color: $text-strong !default;
$label-weight: $weight-bold !default;

$help-size: $size-small !default;

:host {
  .label {
    color: $label-color;
    display: block;
    font-size: $size-normal;
    font-weight: $label-weight;

    &:not(:last-child) {
      margin-bottom: 0.5em;
    }
  }

  .field {
    &:not(:last-child) {
      margin-bottom: 0.75rem;
    }

    // Modifiers
    &.has-addons {
      display: flex;
      justify-content: flex-start;

      .control {
        .button,
        .input,
        .select select {
          &:not([disabled]) {
            &:hover,
            &.is-hovered {
              z-index: 2;
            }

            &:focus,
            &.is-focused,
            &:active,
            &.is-active {
              z-index: 3;

              &:hover {
                z-index: 4;
              }
            }
          }
        }

        &.is-expanded {
          flex-grow: 1;
          flex-shrink: 1;
        }
      }

      &.has-addons-centered {
        justify-content: center;
      }

      &.has-addons-right {
        justify-content: flex-end;
      }

      &.has-addons-fullwidth {
        .control {
          flex-grow: 1;
          flex-shrink: 0;
        }
      }

      .control:not(:last-child) {
        margin-right: -1px;

        &:not(:first-child):not(:last-child) {
          .button,
          .input,
          .select select {
            border-radius: 0;
          }
        }

        &:first-child:not(:only-child) {
          .button,
          .input,
          .select select {
            border-bottom-right-radius: 0;
            border-top-right-radius: 0;
          }
        }

        &:last-child:not(:only-child) {
          .button,
          .input,
          .select select {
            border-bottom-left-radius: 0;
            border-top-left-radius: 0;
          }
        }
      }
    }

    &.is-horizontal {
      @include tablet {
        display: flex;
      }
    }
  }

  ::slotted([slot='hint']) {
    .help {
      display: block;
      font-size: $help-size;
      margin-top: 0.25rem;
      @each $name, $pair in $colors {
        $color: nth($pair, 1);
        &.is-#{$name} {
          color: $color;
        }
      }
    }
  }
}
