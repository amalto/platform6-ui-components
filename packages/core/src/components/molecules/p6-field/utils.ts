import { isP6Control, isP6NativeControl, P6Control, P6NativeControl } from '../../../shared/form/control';
import { isInDefaultSlot } from '../../../utils/component';

// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
export function isControl(elmt: unknown): elmt is P6Control<unknown> | P6NativeControl {
  return isP6Control(elmt) || isP6NativeControl(elmt);
}

export function getControl(children: HTMLCollection): P6Control<unknown> | P6NativeControl | undefined {
  return Array.from(children).filter(isInDefaultSlot).find(isControl) as P6Control<unknown> | P6NativeControl | undefined;
}
