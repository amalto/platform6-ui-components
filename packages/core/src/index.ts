export { Components, JSX } from './components';
export { P6ButtonType } from './components/atoms/p6-button/p6-button';
export { P6CalendarType } from './components/atoms/p6-calendar/p6-calendar';
export { P6CheckboxValue } from './components/atoms/p6-checkbox/p6-checkbox';
export { P6InputValue } from './components/atoms/p6-input/p6-input';
export { P6InputType } from './components/atoms/p6-input/types';
export { Target } from './components/atoms/p6-link/p6-link';
export { P6RadioValue } from './components/atoms/p6-radio/p6-radio';
export { P6SelectControl, P6SelectValue } from './components/atoms/p6-select-native/p6-select-native';
export { P6SwitchValue } from './components/atoms/p6-switch/p6-switch';
export { P6TextareaValue } from './components/atoms/p6-textarea/p6-textarea';
export { P6LanguageValue } from './components/molecules/p6-language/p6-language';
export { P6TranslationControl, P6TranslationValue } from './components/molecules/p6-translation/p6-translation';
export {
  AlignColumnDetail,
  CellValueChangedDetail,
  Column,
  ColumnDefinition,
  ColumnId,
  DataItem,
  EditingCellDetail,
  FilterRowsDetail,
  MoveColumnDetail,
  ResetDefinitionsDetail,
  ResizeColumnDetail,
  Row,
  RowId,
  ShowColumnDetail,
  ShowOptionsDetail,
  SortColumnDetail,
} from './components/organisms/p6-grid/core/entities';
export { ServiceStatus } from './components/templates/p6-service-panel/service-status.enum';
export { P6Control } from './shared/form/control';
export { InvalidEventDetail, isCustomEvent, isInvalidEvent, isValidEvent, ValidEventDetail } from './shared/form/event';
export { Alignment, Mode, modes, Position, positions, Size, sizes, SortOrder } from './shared/types';
export { EnumArrayEntry, enumArrayToObject, enumToArray, EnumType } from './utils/enum';
export { LanguageCode } from './utils/language';
