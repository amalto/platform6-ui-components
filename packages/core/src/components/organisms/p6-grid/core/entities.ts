import { EventEmitter, JSX } from '@stencil/core';
import { Alignment, Direction, SortOrder } from '../../../../shared/types';

export type ColumnId = string;

export type DataItem = Record<string, unknown>;

export interface ColumnDefinition<T extends DataItem> {
  id?: ColumnId;
  label: string;
  field?: string | keyof T;
  editable?: boolean;
  filtreable?: boolean;
  sortable?: boolean;
  hidden?: boolean;
  width?: number;
  align?: Alignment;
  color?: string;
  sortOrder?: SortOrder;
  disableHeaderMenu?: boolean;
  getValue?: (data: T, context: ColumnDefinition<T>) => string;
  setValue?: (value: string, data: T, context: ColumnDefinition<T>) => T;
  cellEditor?: (data: T, context: ColumnDefinition<T>) => JSX.Element & CellEditor<T>;
  cellRenderer?: (data: T, rowId: RowId, context: ColumnDefinition<T>) => string | HTMLElement;
}

export interface Column<T extends DataItem> extends ColumnDefinition<T> {
  id: ColumnId;
  editable: boolean;
  hidden: boolean;
  width: number;
  align: Alignment;
  sortOrder: SortOrder;
  disableHeaderMenu: boolean;
  getValue: (data: T, context: ColumnDefinition<T>) => string;
  setValue: (value: string, data: T, context: ColumnDefinition<T>) => T;
}

export interface CellEditor<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  value: string;

  p6GridCellValueChanged: EventEmitter<CellValueChangedDetail<T>>;
  p6GridEditingCell: EventEmitter<EditingCellDetail<T>>;
}
export type RowId = string;

export interface Row<T extends DataItem> {
  id: RowId;
  data: T;
}

interface ColumnEventDetail<T extends DataItem> {
  column: Column<T>;
}

export interface MoveColumnDetail<T extends DataItem> extends ColumnEventDetail<T> {
  direction: Direction;
}
export type ResizeColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type AlignColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type SortColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type ShowColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export interface CellValueChangedDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  oldValue: string;
}

export enum EditingCellStatus {
  Start,
  Stop,
  Cancel,
}

export interface EditingCellDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  status: EditingCellStatus;
}

export interface ResetDefinitionsDetail {
  reset: boolean;
}

export interface ShowOptionsDetail {
  visible: boolean;
}

export interface FilterRowsDetail {
  value: string;
}
