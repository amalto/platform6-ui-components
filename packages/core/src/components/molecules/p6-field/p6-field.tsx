import { Component, ComponentInterface, Element, h, Host, JSX, State } from '@stencil/core';
import { isP6NativeControl, P6Control, P6NativeControl } from '../../../shared/form/control';
import { isInvalidEvent, isValidEvent } from '../../../shared/form/event';
import { Mode } from '../../../shared/types';
import { isEmpty } from '../../../utils/attribute';
import { getControl } from './utils';

@Component({
  tag: 'p6-field',
  styleUrl: 'p6-field.scss',
  shadow: true,
})
export class P6Field implements ComponentInterface {
  @Element() host!: HTMLP6FieldElement;

  @State() errorMessage = '';

  private input: P6Control<unknown> | P6NativeControl | undefined;

  componentWillLoad(): void {
    this.input = getControl(this.host.children);

    if (this.input !== undefined && isP6NativeControl(this.input)) {
      this.host.addEventListener('focusout', this.updateValidity);
      this.input.addEventListener('invalid', this.invalidNativeHandler);
      if (!isEmpty(this.input?.value)) {
        this.updateValidity();
      }
    }
  }

  componentWillRender(): void {
    const errorHintClassName = 'field-auto_hint_message';
    const errorHintMsg = this.host.querySelector(`.${errorHintClassName}`);

    if (errorHintMsg !== null) {
      this.host.removeChild(errorHintMsg);
    }

    if (this.errorMessage !== '' && !this.isReadOnly) {
      const hint = document.createElement('p6-hint');
      hint.innerText = this.errorMessage;
      hint.mode = Mode.danger;
      hint.slot = 'hint';
      hint.classList.add(errorHintClassName);
      this.host.appendChild(hint);
    }
  }

  render(): JSX.Element {
    return (
      <Host aria-disabled={this.isDisabled ? 'true' : null} onP6Valid={this.validHandler} onP6Invalid={this.invalidHandler}>
        {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
        <label
          class={{
            label: true,
          }}
          htmlFor={this.input?.name}
        >
          <slot name="label" />
          <slot />
        </label>
        {this.isReadOnly || this.isDisabled ? null : <slot name="hint" />}
      </Host>
    );
  }

  disconnectedCallback(): void {
    if (this.input === undefined) {
      return;
    }

    if (isP6NativeControl(this.input)) {
      this.host.removeEventListener('focusout', this.updateValidity);
    }
  }

  private get isReadOnly(): boolean {
    return this.input?.readOnly || false;
  }

  private get isDisabled(): boolean {
    return this.input?.disabled || false;
  }

  private updateValidity = (): void => {
    if (isP6NativeControl(this.input) && this.input?.checkValidity()) {
      this.errorMessage = '';
    }
  };

  private validHandler = (event: Event): void => {
    if (isValidEvent(event)) {
      this.errorMessage = '';
    }
  };

  private invalidHandler = (event: Event): void => {
    if (isInvalidEvent(event)) {
      this.errorMessage = event.detail.message;
    }
  };

  private invalidNativeHandler = (event: Event): void => {
    if (event.target !== null && 'validationMessage' in event.target) {
      const { validationMessage } = event.target;
      this.errorMessage = validationMessage;
    }
  };
}
