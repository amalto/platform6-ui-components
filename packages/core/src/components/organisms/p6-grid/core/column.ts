import { Alignment, Direction, SortOrder } from '../../../../shared/types';
import { throwError } from '../../../../utils/error';
import { Column, ColumnDefinition, ColumnId, DataItem } from './entities';

export const MIN_WIDTH = 75;
export const DEFAULT_WIDTH = MIN_WIDTH;
export const INC_WIDTH = 10;

export function defaultGetValue<T extends DataItem>(data: T, context: ColumnDefinition<T>): string {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'getValue' function for ${JSON.stringify(context)}`);
  }

  return `${data[context.field as keyof T]}`;
}

export function defaultSetValue<T extends DataItem>(value: string, data: T, context: ColumnDefinition<T>): T {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'setValue' function for ${JSON.stringify(context)}`);
  }

  return {
    ...data,
    [context.field as keyof T]: value,
  };
}

let columnIds = 0;
export function fromDefinition<T extends DataItem>(column: ColumnDefinition<T>): Column<T> {
  // eslint-disable-next-line no-plusplus
  const id = column.id || column.field || columnIds++;

  return {
    ...column,
    id: `col-${id}`,
    editable: column.editable || false,
    hidden: column.hidden || false,
    width: column.width || 100,
    align: column.align || Alignment.center,
    sortOrder: column.sortOrder || SortOrder.none,
    disableHeaderMenu: column.disableHeaderMenu || false,
    getValue: column.getValue || defaultGetValue,
    setValue: column.setValue || defaultSetValue,
  };
}

export function isHidden<T extends DataItem>(column: Column<T>): boolean {
  return !!column.hidden;
}

export function replaceColumn<T extends DataItem>(updatedColumn: Column<T>, allColumns: Column<T>[]): Column<T>[] {
  return allColumns.map(col => (col.id === updatedColumn.id ? updatedColumn : col));
}

export function move<T extends DataItem>(id: ColumnId, direction: Direction, columns: Column<T>[]): Column<T>[] {
  const from = columns.findIndex(col => col.id === id);
  const to = from + (direction === Direction.left ? -1 : 1);

  if (from < 0 || to < 0 || to >= columns.length) {
    return columns;
  }

  const result = columns.slice();
  const item = result.splice(from, 1);

  return result.slice(0, to).concat(item).concat(result.slice(to, columns.length));
}
