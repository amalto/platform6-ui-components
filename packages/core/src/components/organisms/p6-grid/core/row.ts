import { SortOrder } from '../../../../shared/types';
import { Column, DataItem, Row, RowId } from './entities';

let rowIds = 0;
export function fromData<T extends DataItem>(data: T): Row<T> {
  rowIds += 1;

  return {
    id: `row-${rowIds}`,
    data,
  };
}

export function compareRow<T extends DataItem>(column: Column<T> | undefined): (a: Row<T>, b: Row<T>) => number {
  const sortOrder = column?.sortOrder || SortOrder.none;
  const factor = sortOrder === SortOrder.desc ? -1 : 1;

  return (a: Row<T>, b: Row<T>): number => {
    if (column === undefined || column.sortable === false || sortOrder === SortOrder.none) {
      return 0;
    }

    const valueA = column.getValue(a.data, column);
    const valueB = column.getValue(b.data, column);

    return valueA.localeCompare(valueB) * factor;
  };
}

export function filterBy<T extends DataItem>(searchValue: string, columns: Column<T>[]): (row: Row<T>) => boolean {
  return row => columns.filter(col => col.filtreable !== false).some(col => col.getValue(row.data, col).includes(searchValue));
}

export function rangeSelectRow<T extends DataItem>(currentRowIdSelect: RowId, selectedRows: Set<RowId>, displayedRow: Row<T>[]): Set<RowId> {
  const currentSelectedIndex = displayedRow.findIndex(row => row.id === currentRowIdSelect);

  if (currentSelectedIndex === -1) {
    return selectedRows;
  }

  const previousRowIdSelected = Array.from(selectedRows)
    .reverse()
    .find(rowId => displayedRow.findIndex(row => row.id === rowId) > -1);

  if (previousRowIdSelected === undefined) {
    return new Set(selectedRows.add(currentRowIdSelect));
  }

  const previousSelectedIndex = displayedRow.findIndex(row => row.id === previousRowIdSelected);

  let [from, to] = [previousSelectedIndex, currentSelectedIndex];
  let excludedFromSlice = currentRowIdSelect;
  if (currentSelectedIndex < previousSelectedIndex) {
    [from, to] = [to, from];
    excludedFromSlice = previousRowIdSelected;
  }

  return displayedRow
    .slice(from, to)
    .reduce((selected, cur) => {
      return cur.id === undefined ? selected : selected.add(cur.id);
    }, new Set(selectedRows))
    .add(excludedFromSlice);
}

export function replaceRow<T extends DataItem>(updatedRow: Row<T>, allRows: Row<T>[]): Row<T>[] {
  return allRows.map(row => (row.id === updatedRow.id ? updatedRow : row));
}
