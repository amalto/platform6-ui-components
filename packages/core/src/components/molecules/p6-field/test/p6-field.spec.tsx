import { newSpecPage } from '@stencil/core/testing';
import { P6Field } from '../p6-field';

describe('p6-field', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [P6Field],
      html: `<p6-field></p6-field>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
