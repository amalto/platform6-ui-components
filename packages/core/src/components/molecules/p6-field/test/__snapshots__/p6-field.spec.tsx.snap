// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-field renders 1`] = `
<p6-field>
  <mock:shadow-root>
    <label class="label">
      <slot name="label"></slot>
      <slot></slot>
    </label>
    <slot name="hint"></slot>
  </mock:shadow-root>
</p6-field>
`;
